### DISCUSSION TITLE
optimize using statement

### DISCUSSION THREAD

**Discussion by Jaime-zhang** [dotnet/csharplang#5467] [Language Ideas]
c# will not optimize it, although the GetTest function always returns null in release version
```
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static IDisposable GetTest()
{
#if DEBUG
   return new Test();
#else
    return null;
#endif
}

public static void MethodA()
{
        {
            using IDisposable filestream = GetTest();
            //code
        }
}
public static void MethodB()
{
        {
            using IDisposable filestream = GetTest();
            //code
        }
}
```

.method private hidebysig static void  MethodA() cil managed
{
  .maxstack  1
  .locals init ([0] class [mscorlib]System.IDisposable filestream)
  IL_0000:  call       class [mscorlib]System.IDisposable ConsoleApp2.Program::GetTest()
  IL_0005:  stloc.0
  .try
  {
    IL_0006:  leave.s    IL_0012
  }  // end .try
  finally
  {
    IL_0008:  ldloc.0
    IL_0009:  brfalse.s  IL_0011
    IL_000b:  ldloc.0
    IL_000c:  callvirt   instance void [mscorlib]System.IDisposable::Dispose()
    IL_0011:  endfinally
  }  // end handler
  IL_0012:  ret
} // end of method Program::MethodA

**Comment by canton7** [comment]
Interestingly, the JIT does optimize this, but only if the `Console.WriteLine` is commented out [here](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdAEoCuMCUYAplQMID2YADlADYsATgGVhANygBjFigDcRIpgDMFbGQAqshGRBkAkgBEoKHpxQRggogG8iZB2QDaAWRYIAFpwAm+3vwAKN08fPx5+AHkeJk4YFCoAQQBzJKFZFChxFn0YflhYJIBKAF17RxUKXAA2A2NTc0tBMgBxdy0UBADCsoc7QgBiKAAzMkMAUQAhAFVmnscKAHYyGBYAd01tLoUBln4UFjnypZg6fn5t/pYYb2G5gF9FQnmKyhrMABYyYK9vBK65vrzZ7kIwmMwWKwsMhDATaNIQMBkAC8LTam0K2yB5VwAE4AgAiABinE4+Ix9zmL0+dXBLC6ZBsZAehDuQA==)

**Comment by 333fred** [comment]
The C# compiler is not an optimizing compiler. We rely on the JIT to inline methods, and do not do anything beyond local slot optimizations. This is very intentional: if we implement optimizations in the compilers for individual languages, that means that every optimization needs to be implemented in every language. By just having the JIT optimize these things, all languages get benefits without having to pay the price themselves.
